File: Design-Patterns-Using-Ruby/Decorator/DecoratorPattern.rb
# Component
class Component
    def operation
    end
end

# ConcreteComponent
class ConcreteComponent < Component
    def operation
        puts 'ConcreteComponent operation.'
    end
end

# Decorator
class Decorator < Component
    def initialize(component)
        @component = component
    end
  
    def operation
        @component.operation
    end
end

# ConcreteDecorator1
class ConcreteDecorator1 < Decorator
    def operation
        puts 'ConcreteDecorator1 operation start.'
        super
        puts 'ConcreteDecorator1 operation end.'
    end
end

# ConcreteDecorator2
class ConcreteDecorator2 < Decorator
    def operation
        puts 'ConcreteDecorator2 operation start.'
        super
        puts 'ConcreteDecorator2 operation end.'
    end
end

# Client code
component = ConcreteComponent.new
decorator1 = ConcreteDecorator1.new(component)
decorator2 = ConcreteDecorator2.new(decorator1)
decorator2.operation

=begin 
ConcreteDecorator2 operation start.
ConcreteDecorator1 operation start.
ConcreteComponent operation.
ConcreteDecorator1 operation end.
ConcreteDecorator2 operation end. 
=end

File: Design-Patterns-Using-Swift/Decorator/DecoratorPattern.swift
// Component
protocol Component {
    func operation()
}

// ConcreteComponent
class ConcreteComponent: Component {
    func operation() {
        print("ConcreteComponent operation.")
    }
}

// Decorator
class Decorator: Component {
    private let component: Component

    init(_ component: Component) {
        self.component = component
    }

    func operation() {
        component.operation()
    }
}

// ConcreteDecorator1
class ConcreteDecorator1: Decorator {
    override func operation() {
        print("ConcreteDecorator1 operation start.")
        super.operation()
        print("ConcreteDecorator1 operation end.")
    }
}

// ConcreteDecorator2
class ConcreteDecorator2: Decorator {
    override func operation() {
        print("ConcreteDecorator2 operation start.")
        super.operation()
        print("ConcreteDecorator2 operation end.")
    }
}

// Client code
let component: Component = ConcreteComponent()
let decorator1: Decorator = ConcreteDecorator1(component)
let decorator2: Decorator = ConcreteDecorator2(decorator1)
decorator2.operation()

/*
ConcreteDecorator2 operation start.
ConcreteDecorator1 operation start.
ConcreteComponent operation.
ConcreteDecorator1 operation end.
ConcreteDecorator2 operation end.
*/